Fix aerodynamic torque simulations

BenChung opened the pull request:
This pull request fixes Ferram Aerospace's simulations for torque and force, as well as adding new API endpoints that allow access to real-time force and torque information. Additionally, it refactors the CalculateAeroForces system to reduce code duplication, though it could be refactored further to avoid the lambda and associated generic method invocation overhead.

BenChung commented:
@dkavolis Should I open issues for the problems that this PR fixes? I realize in retrospect that I'm fixing one issue and adding a new feature (current aerodynamic torque) in one PR, which may not work well for you.

dkavolis commented:
Thanks for the PR, sorry it took so long to have a look at it.
Separate PRs would make more sense, I just have a few comments:
1. In `FarCenterQuery` I would rename `pos` to `origin`, keep `GetMinTorquePos` and `TorqueClipFactor` (even though they are unused, they might be useful in the future so might as well) and modify `TorqueAt` to compensate for non-zero `origin`. I would also provide a default constructor which sets `origin` to zero.
1. In `PhysicsCalcs.cs` the `totalAeroForceVector` and `totalAeroTorqueVector` could be combined into a single `FarCenterQuery`, maybe add a `Reset(Vector3d origin)` method to it and call it before iterating.
1. In `FARAPI.cs` a lot of the methods have counterparts for active vessel, I would add similar ones. Also, all methods should handle cases when `VesselFlightInfo` returns a null instead of throwing an NRE. Something like `VesselFlightInfo(v)?.InfoParameters.aerodynamicForce ?? 0;`. `0` default is in line with other `FARAPI` methods. 
1. In `FARAeroSection.cs` while the changes make code easier to maintain, the anonymous functions also generate garbage. It may not be important in the editor but I would like to keep the generated garbage to a minimum in flight mode for performance reasons.

BenChung commented:
> In FarCenterQuery I would keep[...] GetMinTorquePos and TorqueClipFactor (even though they are unused, they might be useful in the future so might as well) 

Both GetMinTorquePos and TorqueClipFactor were essentially irrevocably broken due to issues inherent to FARCenterQuery, which assumed that p1 x f1 + p2 x f2 - o x (f1 + f2) =p1 x (f1 - o) + p2 x (f2 - o); basically every number FARCenterQuery produced was nonsense. To add them back would require a new approach to their calculation.

>  I would also provide a default constructor which sets origin to zero.

I'm not sure this makes sense; FARCenterQuery is fundamentally asking what the force is around a given point. It makes more sense to me to require that that point be specified rather than setting it to 0 in whatever frame of reference you happen to be using. This is more a design issue, however. 

> In FARAeroSection.cs while the changes make code easier to maintain, the anonymous functions also generate garbage.

Abstracting over the means of force application is going to require a level of indirection, and it's not clear to me how to do it without allocation. The alternatives that I can come up with are:

* Adding two classes and an interface that provides a consistent interface for force application by AeroSections.
* Accumulating applied forces into a buffer for application by the caller.

I'll make the other changes soon.

dkavolis commented:
> I'm not sure this makes sense; FARCenterQuery is fundamentally asking what the force is around a given point. It makes more sense to me to require that that point be specified rather than setting it to 0 in whatever frame of reference you happen to be using. This is more a design issue, however.

With the `TorqueAt` method the default constructor wouldn't make a difference, only a convenient way to initialize `FARCenterQuery`

> Abstracting over the means of force application is going to require a level of indirection, and it's not clear to me how to do it without allocation. The alternatives that I can come up with are:
>
>   * Adding two classes and an interface that provides a consistent interface for force application by AeroSections.
>   * Accumulating applied forces into a buffer for application by the caller.

I think the first alternative would be the easiest and most extendable in the future if needed

BenChung commented:
I worked on it more, and FARCenterQuery was actually correct despite what I thought at the time; I've reverted the changes locally. 

The bad news is that I've been validating it again and the simulation forces are still broken; I thought I had fixed it, but apparently not. Going to be a bit longer, sadly.

BenChung commented:
Update: torque is now applied consistently in the right direction, and total forces are consistently within 10% of actual values. Going to try to get this lower still though.

BenChung commented:
The remaining 10% of error is due to issues in calculating local atmospheric density due to not knowing the time of day and therefore the solar-induced atmospheric heating. I'm going to call it good here, and address the usability/stability concerns from here on out.

BenChung commented:
I think that's everything; @dkavolis can you please take a look again?

dkavolis commented:
Mostly good, just a couple of comments:
1. The new API calls should have similarly named methods for active vessel that take no arguments to make them consistent with other vessel information getters
1. Is there a particular reason why `FARCenterQuery` and contexts are initialized to `null`? Why not just initialize them once in the constructor (or constructor equivalent) and skip null checks altogether?

BenChung commented:
@dkavolis I think that's everything?

> Mostly good, just a couple of comments:
> 
> 1. The new API calls should have similarly named methods for active vessel that take no arguments to make them consistent with other vessel information getters

Yep, done, apologies for forgetting this originally.

> 2. Is there a particular reason why `FARCenterQuery` and contexts are initialized to `null`? Why not just initialize them once in the constructor (or constructor equivalent) and skip null checks altogether?

It didn't seem to work for mysterious reasons the first time I tried it. I did an inline initalizer and it's fine?

dkavolis commented on FerramAerospaceResearch/FARAPI.cs:
```diff
@@ -401,6 +401,24 @@ public static Vector3 VesselAerodynamicTorque(Vessel v)
+        /// Returns the current aerodynamic force being experienced by the active vehicle in world space
+        /// </summary>
+        /// <returns>The force on the vessel in world space</returns>
+        public static Vector3 ActiveVesselAerodynamicForce(Vessel v)
```
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/ec14134c56a450569cf5c4fcf2b00734d2b7aaaa/FerramAerospaceResearch/FARAPI.cs#L409
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/ec14134c56a450569cf5c4fcf2b00734d2b7aaaa/FerramAerospaceResearch/FARAPI.cs#L418

No need for `Vessel` arguments

dkavolis commented on FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:
```diff
@@ -418,40 +507,26 @@ public void ClearAeroSection()
+            if (simContext == null)
+                simContext = new SimulatedForceContext(vel, center, atmDensity);
+            else
+                simContext.UpdateSimulationContext(vel, center, atmDensity);
```
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/3459b30b70fda45e9573da643b0bcfffc3e427c2/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs#L517-L520
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/3459b30b70fda45e9573da643b0bcfffc3e427c2/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs#L527-L528

Same as with `FARCenterQuery`, initialize in ctor and skip null checks

BenChung commented:
Apologies, missed those; is that sufficient?

dkavolis commented on FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:
```diff
@@ -285,182 +286,103 @@ public void ClearAeroSection()
-
-                moment *= momentFactor;
-                dampingMoment *= momentFactor;
+                return pd.aeroModule.part.partTransform.InverseTransformVector(worldVel);
```
This throws NREs when the vehicle breaks up (if MechJeb is installed). Maybe add another method (something like `IsValid(PartData pd)`?) to the interface which checks if the current `PartData pd` is valid. So for flight forces it would be `(object)pd.aeroModule != null` and for simulation in addition check if `pd.aeroModule.part != null` and `pd.aeroModule.part.partTransform != null`.

Also, consider caching the return value since `Transform.InverseTransformVector` calls an internal method (managed to native) which is slow (compared to usual C# calls).

BenChung replied:
There's already a check `(object)pd.aeroModule != null` in the calling context; see line 389 of FARAeroSection.cs. I think it's okay if simulation produces a 0 for that part in the latter two cases (and I'm not overly concerned with simulation performance; I can basically do a full wind tunnel like test in <0.5s), so to minimize vcalls in the flight code I'll put that into SimulationContext only.

dkavolis commented on FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:
```diff
@@ -336,12 +337,15 @@ public void UpdateSimulationContext(Vector4 worldVel, FARCenterQuery center, flo
             public Vector3 LocalVelocity(PartData pd)
             {
+                if (pd.aeroModule.part == null || pd.aeroModule.part.partTransform == null)
+                    return Vector3.zero;
```
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/593cf15f27bbd95f3a602c690cb795d5d956d53c/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs#L340-L341
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/593cf15f27bbd95f3a602c690cb795d5d956d53c/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs#L347-L348

Rather than doing the same null check twice, do it once and skip the iteration if it is not satisfied. There is no need to waste computation on invalid parts.

BenChung replied:
In this circumstance, should the non-simulation path also skip? It would be easiest just to lift these checks into CalculateAeroForces.

dkavolis replied:
In flight calculation mode, the null checks are not needed and I would like it to stay that way. Unity `Object` overrides equality operators with a call to native code so avoiding them in methods that are called frequently helps with performance.

BenChung replied:
If this is performance critical (I'm still not sure that the simulation path is?), then I'm unhappy introducing another vcall if it can possibly be avoided. Another option would be to skip if the local velocity is exactly zero. In that case, the applied forces would then also be zero, so it shouldn't change the net result.

dkavolis replied:
MechJeb calls simulation from `FixedUpdate`. If you think checking if local velocity, or rather its `sqrMagnitude`, is zero is better than additional interface method call, then go ahead and do that. There's an extension method `FARMathUtil.NearlyEqual` for approximate floating point equality, negligible velocities could also be discarded this way.

dkavolis commented on FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:
```diff
@@ -285,182 +286,107 @@ public void ClearAeroSection()
-                    momentFactor = 0.6f * hypersonicMomentForward;
-                else
+                if (pd.aeroModule.part == null || pd.aeroModule.part.partTransform == null)
+                    return;
```
This is now redundant.

BenChung replied:
Thanks, forgot that I had added that.

dkavolis commented:
Looks like everything is done now, thanks.
